components/api_request_helper/api_request_helper.cc

```diff
@@ -335,106 +314,177 @@ void APIRequestHelper::URLLoaderHandler::send_sse_data_for_testing(
+  // we don't delete the reference to |data_decoder_| which would cancel the
+  // operations.
+  // Response must be complete to know that decoding is definitely complete
+  // since streaming responses may get more reponse chunks which need decoding.
+  const bool decoding_is_complete = (current_decoding_operation_count_ == 0);
+
+  if (request_is_finished_ && decoding_is_complete) {
+    std::move(result_callback_).Run(ToAPIRequestResult(std::move(url_loader_)));
+  } else {
+    VLOG(3) << "Did not run URLLoaderHandler completion handler, still have "
+            << current_decoding_operation_count_
+            << " decoding operations in progress, waiting for them to"
+            << " complete...";
+  }
 }
 
 void APIRequestHelper::URLLoaderHandler::ParseSSE(
     base::StringPiece string_piece) {
   // We split the string into multiple chunks because there are cases where
   // multiple chunks are received in a single call.
-  std::vector<std::string> stream_data = base::SplitString(
+  std::vector<base::StringPiece> stream_data = base::SplitStringPiece(
       string_piece, "\r\n", base::KEEP_WHITESPACE, base::SPLIT_WANT_NONEMPTY);
 
   // TODO(@nullhook): Identify both JSON and string values. The below currently
   // only identifies JSON values.
   static constexpr char kDataPrefix[] = "data: {";
 
   for (const auto& data : stream_data) {
+    DVLOG(3) << "Received chunk: " << data;
     if (!base::StartsWith(data, kDataPrefix)) {
+      // This is useful to know in case an API starts coming back
+      // with unknown data type in some scenarios.
+      VLOG(1) << "Data did not start with SSE prefix";
       continue;
     }
 
-    std::string json = data.substr(strlen(kDataPrefix) - 1);
-
+    auto json = data.substr(strlen(kDataPrefix) - 1);
     auto on_json_parsed =
         [](base::WeakPtr<APIRequestHelper::URLLoaderHandler> handler,
            data_decoder::DataDecoder::ValueOrError result) {
+          DVLOG(2) << " Chunk parsed";
           if (!handler) {
             return;
           }
+          handler->current_decoding_operation_count_--;
           DCHECK(handler->data_received_callback_);
           handler->data_received_callback_.Run(std::move(result));
+          // Parsing is potentially the last operation for |URLLoaderHandler|.
+          handler->MaybeSendResult();
         };
-
-    api_request_helper_->data_decoder_->ParseJson(
-        json, base::BindOnce(on_json_parsed, weak_ptr_factory_.GetWeakPtr()));
+    // Keep track of number of in-progress data decoding operations
+    // so that we can know if any are still in-progress when the request
+    // completes.
+    DVLOG(2) << "Going to call ParseJson";
+    current_decoding_operation_count_++;
+    GetDataDecoder()->ParseJson(std::move(json.data()),
+                                base::BindOnce(std::move(on_json_parsed),
+                                               weak_ptr_factory_.GetWeakPtr()));
```

Comment:
There is now guarantee that ParseJson always(and will always) calls on_json_parsed asynchronously, so after current_decoding_operation_count_ decrement it can become zero while outer loop is still running. 
Let's prematurely calculate current_decoding_operation_count_ before loop starts.

Reply:
0 is ok, right? If the request is not completed and `request_is_finished_` is still `false` then it won't end the handling. And if it's Synchronous, then the completed event cannot happen before on_json_parsed` is called as there is nothing else Asynchronous within the loop. But if I'm mistaken or it makes it clearer code for you then I can try to count the items before running the parser.

Reply:
did your suggestion here: https://github.com/brave/brave-core/compare/41c9c0755d2541fef6c9e838b164079889ec5cfe..5bd56e833219197a2314ea486f423a0e81b28d40